File: mvvmhabit/src/main/java/me/goldze/mvvmhabit/binding/viewadapter/searchview/ViewAdapter.java
package me.goldze.mvvmhabit.binding.viewadapter.searchview;

import android.text.TextUtils;
import android.view.View;

import androidx.databinding.BindingAdapter;

import com.jakewharton.rxbinding2.widget.RxTextView;
import com.mancj.materialsearchbar.MaterialSearchBar;
import com.mancj.materialsearchbar.adapter.SuggestionsAdapter;

import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import me.goldze.mvvmhabit.binding.command.BindingCommand;

/**
 * @author Alwyn
 * @Date 2020/11/4
 * @Description
 */
public class ViewAdapter {

    @BindingAdapter(value = {"onNavigationCommand", "onSearchConfirmCommand","onSearchStateCommand"}, requireAll = false)
    public static void onSearchDrawerClickCommand(MaterialSearchBar searchBar,
                                                  BindingCommand<Void> command,
                                                  BindingCommand<String> bindingCommand,
                                                  BindingCommand<Boolean> booleanCommand) {
        searchBar.setOnSearchActionListener(new MaterialSearchBar.OnSearchActionListener() {
            @Override
            public void onSearchStateChanged(boolean enabled) {
                if (booleanCommand!=null){
                    booleanCommand.execute(enabled);
                }
            }

            @Override
            public void onSearchConfirmed(CharSequence text) {
                if (bindingCommand != null && !TextUtils.isEmpty(text)) {
                    bindingCommand.execute(text.toString());
                }
            }

            @Override
            public void onButtonClicked(int buttonCode) {
                if (MaterialSearchBar.BUTTON_NAVIGATION == buttonCode && command != null) {
                    command.execute();
                }
            }
        });
    }

    @BindingAdapter("onSuggestionItemCommand")
    public static void onSuggestionClick(MaterialSearchBar searchBar, BindingCommand<Integer> command) {
        searchBar.setSuggestionsClickListener(new SuggestionsAdapter.OnItemViewClickListener() {
            @Override
            public void OnItemClickListener(int position, View v) {
                if (command != null) {
                    command.execute(position);
                }
            }

            @Override
            public void OnItemDeleteListener(int position, View v) {

            }
        });
    }

    @BindingAdapter("onSearchTextChangeCommand")
    public static void onSearchTextChangeCommand(MaterialSearchBar searchBar, BindingCommand<String> command) {
        RxTextView.textChanges(searchBar.getSearchEditText())
                .skip(1)
                .debounce(1L, TimeUnit.SECONDS)
                .map(x -> x.toString())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(x -> {
                    if (command != null) {
                        command.execute(x);
                    }
                });
    }
}

File: lib_base/src/main/java/com/czl/lib_base/binding/viewadapter/listview/ViewAdapter.kt
package com.czl.lib_base.binding.viewadapter.listview

import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ListView
import androidx.databinding.BindingAdapter
import com.czl.lib_base.binding.command.BindingCommand
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit

object ViewAdapter {
    @JvmStatic
    @BindingAdapter(
        value = ["onScrollChangeCommand", "onScrollStateChangedCommand"],
        requireAll = false
    )
    fun onScrollChangeCommand(
        listView: ListView,
        onScrollChangeCommand: BindingCommand<ListViewScrollDataWrapper?>?,
        onScrollStateChangedCommand: BindingCommand<Int?>?
    ) {
        listView.setOnScrollListener(object : AbsListView.OnScrollListener {
            private var scrollState = 0
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {
                this.scrollState = scrollState
                onScrollStateChangedCommand?.execute(scrollState)
            }

            override fun onScroll(
                view: AbsListView,
                firstVisibleItem: Int,
                visibleItemCount: Int,
                totalItemCount: Int
            ) {
                onScrollChangeCommand?.execute(
                    ListViewScrollDataWrapper(
                        scrollState,
                        firstVisibleItem,
                        visibleItemCount,
                        totalItemCount
                    )
                )
            }
        })
    }

    @JvmStatic
    @BindingAdapter(value = ["onItemClickCommand"], requireAll = false)
    fun onItemClickCommand(listView: ListView, onItemClickCommand: BindingCommand<Int?>?) {
        listView.onItemClickListener =
            AdapterView.OnItemClickListener { parent, view, position, id ->
                onItemClickCommand?.execute(position)
            }
    }

    @JvmStatic
    @BindingAdapter("onLoadMoreCommand")
    fun onLoadMoreCommand(listView: ListView, onLoadMoreCommand: BindingCommand<Int?>) {
        listView.setOnScrollListener(OnScrollListener(listView, onLoadMoreCommand))
    }

    class OnScrollListener(listView: ListView, onLoadMoreCommand: BindingCommand<Int?>) :
        AbsListView.OnScrollListener {
        private val methodInvoke = PublishSubject.create<Int>()
        private val onLoadMoreCommand: BindingCommand<Int?>?
        private val listView: ListView
        override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}
        override fun onScroll(
            view: AbsListView,
            firstVisibleItem: Int,
            visibleItemCount: Int,
            totalItemCount: Int
        ) {
            if (firstVisibleItem + visibleItemCount >= totalItemCount && totalItemCount != 0 && (totalItemCount != listView.headerViewsCount
                        + listView.footerViewsCount)
            ) {
                if (onLoadMoreCommand != null) {
                    methodInvoke.onNext(totalItemCount)
                }
            }
        }

        init {
            this.onLoadMoreCommand = onLoadMoreCommand
            this.listView = listView
            methodInvoke.throttleFirst(1, TimeUnit.SECONDS)
                .subscribe { integer -> onLoadMoreCommand.execute(integer) }
        }
    }

    class ListViewScrollDataWrapper(
        var scrollState: Int,
        var firstVisibleItem: Int,
        var visibleItemCount: Int,
        var totalItemCount: Int
    )
}